core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -380,18 +382,25 @@ class ContainerProxy(
     val activationWithLogs: Future[Either[ActivationLogReadingError, WhiskActivation]] = activation
       .flatMap { activation =>
         val start = tid.started(this, LoggingMarkers.INVOKER_COLLECT_LOGS)
-        collectLogs(tid, job.msg.user, activation, container, job.action)
-          .andThen {
-            case Success(_) => tid.finished(this, start)
-            case Failure(t) => tid.failed(this, start, s"reading logs failed: $t")
-          }
-          .map(logs => Right(activation.withLogs(logs)))
-          .recover {
-            case LogCollectingException(logs) =>
-              Left(ActivationLogReadingError(activation.withLogs(logs)))
-            case _ =>
-              Left(ActivationLogReadingError(activation.withLogs(ActivationLogs(Vector(Messages.logFailure)))))
-          }
+        Try(
+          collectLogs(tid, job.msg.user, activation, container, job.action)
+            .andThen {
+              case Success(_) => tid.finished(this, start)
+              case Failure(t) => tid.failed(this, start, s"reading logs failed: $t")
+            }
+            .map(logs => Right(activation.withLogs(logs)))
+            .recover {
+              case LogCollectingException(logs) =>
+                Left(ActivationLogReadingError(activation.withLogs(logs)))
+              case _ =>
+                Left(ActivationLogReadingError(activation.withLogs(ActivationLogs(Vector(Messages.logFailure)))))
```

Comment:
my scala game is not great, but can you remove either the `.recover` or the `case Failure(t)` below? i.e. instead of using Try/Success/Failure, can you just add `tid.failed(this, start, s"reading logs failed: $t")` to the previously existing `.recover case _`?

Reply:
@markusthoemmes can you comment on this when you have a sec?

Reply:
Hmm, no you can't. These errors are related to the inner `Future` while the block below catches the errors from the outer `Try`. I'll try to code something up to fold them both into one.

Reply:
Here it comes, I propose the following diff for this file to narrow down the exception handling:

```diff
diff --git a/core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala b/core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala
index 23d09651..93399069 100644
--- a/core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala
+++ b/core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala
@@ -380,7 +380,11 @@ class ContainerProxy(
     val activationWithLogs: Future[Either[ActivationLogReadingError, WhiskActivation]] = activation
       .flatMap { activation =>
         val start = tid.started(this, LoggingMarkers.INVOKER_COLLECT_LOGS)
-        collectLogs(tid, job.msg.user, activation, container, job.action)
+        val collected: Future[ActivationLogs] = try {
+          collectLogs(tid, job.msg.user, activation, container, job.action)
+        } catch { case t: Throwable => Future.failed(t) }
+
+        collected
           .andThen {
             case Success(_) => tid.finished(this, start)
             case Failure(t) => tid.failed(this, start, s"reading logs failed: $t")

```

It transforms a failure in `collectLogs` to a failed Future, so we can process it as is. WDYT?

Reply:
Much simpler! I was trying to unnecessarily use `Try` where good old `try` would have been better

Reply:
You can always change to `catch { case NonFatal(t) => Future.failed(t) }`